src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -127,15 +132,13 @@ public CompletionStage<SubscriptionResponse> subscribe(UUri topic, UListener lis
                 return CompletableFuture.completedFuture(response);
             })
 
-            // Then Add the handler (if present) to the hashMap
+            // Then Add the handler (if the client provided one) only if we are in a SUBSCRIBE_PENDING state
+            // so that we can be notified when the subscription state changes to SUBSCRIBED.
             .thenApply(response -> {
-                if (handler != null) {
-                    mHandlers.compute(topic, (key, currentHandler) -> {
-                        if (currentHandler != null) {
-                            throw new UStatusException(UCode.ALREADY_EXISTS, "Handler already registered");
-                        }
-                        return handler;
-                    });
+                if  (handler != null && 
+                     response.getStatus().getState() == SubscriptionStatus.State.SUBSCRIBE_PENDING && 
```

Comment:
subsequent subscriptions with the different handler won't trigger this exception, as the state will typically be 'SUBSCRIBED'. Can we do this to achieve similar functionality

` mHandlers.compute(topic, (k, existingHandler) -> {
                if (existingHandler != null && !existingHandler.equals(handler)) {
                    throw new UStatusException(UCode.ALREADY_EXISTS, "Handler already registered");
                }
                return handler;
            });
`

Reply:
done